src/Processors/Sources/MongoDBSource.cpp

```diff
@@ -297,73 +353,77 @@ namespace
-                const auto parse_value = array_info[idx].parser;
-                std::vector<Row> dimensions(expected_dimensions + 1);
-
-                auto array = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(value).value();
+                const auto nested_type = assert_cast<const DataTypeArray &>(data_type).getNestedType();
+                auto parent = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(value).value();
+                const size_t size = parent->size();
 
-                std::vector<std::pair<const Poco::MongoDB::Element *, size_t>> arrays;
-                arrays.emplace_back(&value, 0);
+                Field field = Array();
+                Array & arr = field.get<Array &>();
+                arr.reserve(size);
 
-                while (!arrays.empty())
+                for (size_t child_idx = 0; child_idx < size; ++child_idx)
                 {
-                    size_t dimension_idx = arrays.size() - 1;
+                    Poco::MongoDB::Element::Ptr child = parent->get(static_cast<int>(child_idx));
+                    arr.push_back(getValueField(nested_type, *child, name));
+                }
 
-                    if (dimension_idx + 1 > expected_dimensions)
-                        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Got more dimensions than expected");
+                assert_cast<ColumnArray &>(column).insert(field);
+                break;
+            }
+            case ValueType::vtTuple:
+            {
+                if (value.type() != Poco::MongoDB::ElementTraits<MongoDocument::Ptr>::TypeId)
+                    throw Exception(ErrorCodes::TYPE_MISMATCH, "Type mismatch, expected Tuple, got type id = {} for column {}",
+                                    toString(value.type()), name);
 
-                    auto [parent_ptr, child_idx] = arrays.back();
-                    auto parent = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(*parent_ptr).value();
+                auto document = static_cast<const Poco::MongoDB::ConcreteElement<MongoDocument::Ptr> &>(value).value();
 
-                    if (child_idx >= parent->size())
-                    {
-                        arrays.pop_back();
+                ColumnTuple & column_tuple = assert_cast<ColumnTuple &>(column);
 
-                        if (dimension_idx == 0)
-                            break;
+                const auto * tuple_type = assert_cast<const DataTypeTuple *>(&data_type);
+                const Strings & names = tuple_type->getElementNames();
+                const DataTypes & nested_types = tuple_type->getElements();
 
-                        dimensions[dimension_idx].emplace_back(Array(dimensions[dimension_idx + 1].begin(), dimensions[dimension_idx + 1].end()));
-                        dimensions[dimension_idx + 1].clear();
+                for (const auto col_num : collections::range(0, column_tuple.tupleSize()))
```

Comment:
Just curious shy not just `for (size_t i = 0; i != column_tuple.tupleSize(); ++i)` ?